File: typescript/tests/lr.test.ts
// ===========================================================================
// SECTION 9: LEFT RECURSION (10 tests)
// ===========================================================================

import { testParse } from './testUtils.js';

describe('Left Recursion', () => {
  const lr1 = `
    S <- S "+" T / T ;
    T <- [0-9]+ ;
  `;

  const expr = `
    S <- E ;
    E <- E "+" T / T ;
    T <- T "*" F / F ;
    F <- "(" E ")" / [0-9] ;
  `;

  test('LR01-simple', () => {
    const { ok, errorCount } = testParse(lr1, '1+2+3');
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
  });

  test('LR02-single', () => {
    const { ok, errorCount } = testParse(lr1, '42');
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
  });

  test('LR03-chain 5', () => {
    const { ok, errorCount } = testParse(lr1, Array(5).fill('1').join('+'));
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
  });

  test('LR04-chain 10', () => {
    const { ok, errorCount } = testParse(lr1, Array(10).fill('1').join('+'));
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
  });

  test('LR05-with mult', () => {
    const { ok, errorCount } = testParse(expr, '1+2*3');
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
  });

  test('LR06-parens', () => {
    const { ok, errorCount } = testParse(expr, '(1+2)*3');
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
  });

  test('LR07-nested parens', () => {
    const { ok, errorCount } = testParse(expr, '((1+2))');
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
  });

  test('LR08-direct', () => {
    const { ok, errorCount } = testParse(
      'S <- S "x" / "y" ;',
      'yxxx'
    );
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
  });

  test('LR09-multi-digit', () => {
    const { ok, errorCount } = testParse(lr1, '12+345+6789');
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
  });

  test('LR10-complex expr', () => {
    const { ok, errorCount } = testParse(expr, '1+2*3+4*5');
    expect(ok).toBe(true);
    expect(errorCount).toBe(0);
  });
});

File: dart/test/parser/lr_test.dart
// ===========================================================================
// SECTION 9: LEFT RECURSION (10 tests)
// ===========================================================================

import 'package:test/test.dart';

import '../test_utils.dart';

void main() {
  const lr1 = '''
    S <- S "+" T / T ;
    T <- [0-9]+ ;
  ''';

  test('LR01-simple', () {
    final (ok, err, _) = testParse(lr1, '1+2+3');
    expect(ok, isTrue, reason: 'should succeed');
    expect(err, equals(0), reason: 'should have 0 errors');
  });

  test('LR02-single', () {
    final (ok, err, _) = testParse(lr1, '42');
    expect(ok, isTrue, reason: 'should succeed');
    expect(err, equals(0), reason: 'should have 0 errors');
  });

  test('LR03-chain 5', () {
    final (ok, err, _) = testParse(lr1, List.filled(5, '1').join('+'));
    expect(ok, isTrue, reason: 'should succeed');
    expect(err, equals(0), reason: 'should have 0 errors');
  });

  test('LR04-chain 10', () {
    final (ok, err, _) = testParse(lr1, List.filled(10, '1').join('+'));
    expect(ok, isTrue, reason: 'should succeed');
    expect(err, equals(0), reason: 'should have 0 errors');
  });

  const expr = '''
    S <- E ;
    E <- E "+" T / T ;
    T <- T "*" F / F ;
    F <- "(" E ")" / [0-9] ;
  ''';

  test('LR05-with mult', () {
    final (ok, err, _) = testParse(expr, '1+2*3');
    expect(ok, isTrue, reason: 'should succeed');
    expect(err, equals(0), reason: 'should have 0 errors');
  });

  test('LR06-parens', () {
    final (ok, err, _) = testParse(expr, '(1+2)*3');
    expect(ok, isTrue, reason: 'should succeed');
    expect(err, equals(0), reason: 'should have 0 errors');
  });

  test('LR07-nested parens', () {
    final (ok, err, _) = testParse(expr, '((1+2))');
    expect(ok, isTrue, reason: 'should succeed');
    expect(err, equals(0), reason: 'should have 0 errors');
  });

  test('LR08-direct', () {
    final (ok, err, _) = testParse(
      'S <- S "x" / "y" ;',
      'yxxx',
    );
    expect(ok, isTrue, reason: 'should succeed');
    expect(err, equals(0), reason: 'should have 0 errors');
  });

  test('LR09-multi-digit', () {
    final (ok, err, _) = testParse(lr1, '12+345+6789');
    expect(ok, isTrue, reason: 'should succeed');
    expect(err, equals(0), reason: 'should have 0 errors');
  });

  test('LR10-complex expr', () {
    final (ok, err, _) = testParse(expr, '1+2*3+4*5');
    expect(ok, isTrue, reason: 'should succeed');
    expect(err, equals(0), reason: 'should have 0 errors');
  });
}
